expr/unflattener.go

```diff
@@ -0,0 +1,101 @@
+package expr
+
+import (
+	"github.com/brimsec/zq/field"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/builder"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+type Unflattener struct {
+	zctx        *resolver.Context
+	builders    map[int]*builder.ColumnBuilder
+	recordTypes map[int]*zng.TypeRecord
+	fieldExpr   Evaluator
+}
+
+// NewUnflattener returns a Unflattener that turns successive dotted
+// field names into nested records.  For example, unflattening {"a.a":
+// 1, "a.b": 1} results in {a:{a:1,b:1}}.  Note that while
+// unflattening is applied recursively from the top-level and applies
+// to arbitrary-depth dotted names, it is not applied to dotted names
+// that start at lower levels (for example {a:{"a.a":1}} is
+// unchanged).
+func NewUnflattener(zctx *resolver.Context) *Unflattener {
+	return &Unflattener{
+		zctx:        zctx,
+		builders:    make(map[int]*builder.ColumnBuilder),
+		recordTypes: make(map[int]*zng.TypeRecord),
+	}
+}
+
+func (u *Unflattener) lookupBuilderAndType(in *zng.TypeRecord) (*builder.ColumnBuilder, *zng.TypeRecord, error) {
+	b, ok := u.builders[in.ID()]
+	if ok {
+		b.Reset()
```

Comment:
NIt: I think you should do this in Apply rather than here so it's clear there that the builder is being reset before use.

Reply:
👍 agree